src/mongoose_batch_worker.erl

```diff
@@ -0,0 +1,145 @@
+    ?UNEXPECTED_CALL(Msg, From),
+    {reply, unexpected_call, State}.
+
+-spec handle_cast(task | term(), state()) -> {noreply, state()}.
+handle_cast({task, Task}, State) ->
+    {noreply, handle_task(Task, State)};
+handle_cast(Msg, State) ->
+    ?UNEXPECTED_CAST(Msg),
+    {noreply, State}.
+
+-spec handle_info(flush | term(), state()) -> {noreply, state()}.
+handle_info(flush, State) ->
+    {noreply, run_flush(State)};
+handle_info(Msg, State) ->
+    ?UNEXPECTED_INFO(Msg),
+    {noreply, State}.
+
+-spec terminate(term(), state()) -> term().
+terminate(_Reason, State = #state{flush_queue = []}) ->
+    ?LOG_INFO(#{what => stopping_worker, host_type => State#state.host_type,
+                 flush_callback => State#state.flush_callback}),
+    ok;
+terminate(_Reason, State) ->
+    ?LOG_INFO(#{what => stopping_worker, host_type => State#state.host_type,
+                 flush_callback => State#state.flush_callback}),
+    ?LOG_WARNING(#{what => worker_terminate_requires_flush,
+                   host_type => State#state.host_type,
+                   flush_callback => State#state.flush_callback,
+                   flush_queue_len => length(State#state.flush_queue)}),
+    run_flush(State),
+    ok.
+
+-spec code_change(term(), state(), term()) -> {ok, state()}.
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+% Don't leak the tasks to logs, can contain private information
+format_status(_Opt, [_PDict, State | _]) ->
+    [{data, [{"State", State#state{flush_queue = []}},
+             {"Task Queue Length", length(State#state.flush_queue)}]}].
+
+%%====================================================================
+%% Batched tasks callbacks
+%%====================================================================
+
+handle_task(Task, State0 = #state{batch_size = MaxSize,
+                                  flush_interval = Interval,
+                                  flush_interval_tref = TRef0,
+                                  flush_queue = Acc0}) ->
+    Length = length(Acc0),
+    TRef1 = maybe_schedule_flush(TRef0, Length, Interval),
+    State1 = State0#state{flush_interval_tref = TRef1,
+                          flush_queue = [Task | Acc0]},
+    case Length + 1 >= MaxSize of
+        false -> State1;
+        true -> run_flush(State1)
+    end.
+
+maybe_schedule_flush(undefined, 0, Interval) ->
+    erlang:send_after(Interval, self(), flush);
+maybe_schedule_flush(TRef, _, _) ->
+    TRef.
+
+run_flush(State = #state{flush_queue = Queue, flush_interval_tref = TRef}) ->
+    cancel_and_flush_timer(TRef),
+    ?LOG_DEBUG(#{what => worker_flush, flush_callback => State#state.flush_callback,
+                 host_type => State#state.host_type,
+                 flush_queue_len => length(State#state.flush_queue)}),
+    NewState = do_run_flush(State#state{flush_queue = lists:reverse(Queue),
+                                        flush_interval_tref = undefined}),
+    erlang:garbage_collect(),
```

Comment:
wondering if that would collect `Queue` (something says to me it will still keep Queue, because the caller is still referencing it).
There is another way to collect for gen_server: is to return `{noreply,NewState,hibernate}` after a flush, could be more efficient.

Reply:
hibernate more efficient, not at all, because it will not only GC the heap, but also collapse the whole process memory to a PCB, which will then require a new heap when the process is woken up, which will be a very small heap because hibernation imposes a shrink as well, that will very likely need to be grown once the queue starts filling... no way.

But I'll make sure the queue is GC'ed here, hold on.

Reply:
You were right, it wasn't collected. But got how to, see https://github.com/esl/MongooseIM/pull/3433/commits/ee83696f6adcc7a261e845ead5caf31418ef314c 😁